In [1]:
#Librerias

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf

LECTURA DE DATOS

In [2]:
df_general = pd.read_csv('general_data.csv')

#tamaño de dataframe general
df_general.shape

(8820, 24)

In [3]:
df_encuesta = pd.read_csv('employee_survey_data.csv')

#tamaño de dataframe encuesta
df_encuesta.shape

(8820, 5)

In [4]:
df_manager = pd.read_csv('manager_survey.csv')

#tamaño de dataframe manager
df_manager.shape

(8820, 4)

In [5]:
df_retiro= pd.read_csv('retirement_info.csv')

#tamaño de dataframe retirement
df_retiro.shape

(711, 5)

Respecto a la base de datos general, se puede observar que al analizarla hay algunas columnas que son candidatas a ser eliminadas. Columnas como Over18, revisar StandardHours	para ver si hay datos diferentes a 8, standardhours hay que analizar si tiene valores diferentes a 8.

Tambien es importante destacar que hay columnas que posiblemente presenten una correlación fuerte con nuestra variable objetivo (retiros), como PercentSalaryHike, YearsAtCompany, YearsWithCurrManager

In [6]:
df_general.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8820 entries, 0 to 8819
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      8820 non-null   int64  
 1   BusinessTravel           8820 non-null   object 
 2   Department               8820 non-null   object 
 3   DistanceFromHome         8820 non-null   int64  
 4   Education                8820 non-null   int64  
 5   EducationField           8820 non-null   object 
 6   EmployeeCount            8820 non-null   int64  
 7   EmployeeID               8820 non-null   int64  
 8   Gender                   8820 non-null   object 
 9   JobLevel                 8820 non-null   int64  
 10  JobRole                  8820 non-null   object 
 11  MaritalStatus            8820 non-null   object 
 12  MonthlyIncome            8820 non-null   int64  
 13  NumCompaniesWorked       8782 non-null   float64
 14  Over18                  

In [7]:
#Renombrar columnna de fecha que aparece como InfoDate a Year
df_general = df_general.rename(columns={'InfoDate': 'Year'})

In [7]:
#revision de columnas
df_general

Age     BusinessTravel              Department  DistanceFromHome  \
0      51      Travel_Rarely                   Sales                 6   
1      31  Travel_Frequently  Research & Development                10   
2      32  Travel_Frequently  Research & Development                17   
3      38         Non-Travel  Research & Development                 2   
4      32      Travel_Rarely  Research & Development                10   
...   ...                ...                     ...               ...   
8815   42      Travel_Rarely  Research & Development                 5   
8816   29      Travel_Rarely  Research & Development                 2   
8817   25      Travel_Rarely  Research & Development                25   
8818   42      Travel_Rarely                   Sales                18   
8819   40      Travel_Rarely  Research & Development                28   

      Education EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
0             2  Life Sciences              1           1  Female         1   
1             1  Life Sciences              1           2  Female         1   
2             4          Other              1           3    Male         4   
3             5  Life Sciences              1           4    Male         3   
4             1        Medical              1           5    Male         1   
...         ...            ...            ...         ...     ...       ...   
8815          4        Medical              1        4406  Female         1   
8816          4        Medical              1        4407    Male         1   
8817          2  Life Sciences              1        4408    Male         2   
8818          2        Medical              1        4409    Male         1   
8819          3        Medical              1        4410    Male         2   

      ... Over18 PercentSalaryHike  StandardHours  StockOptionLevel  \
0     ...      Y                11              8                 0   
1     ...      Y                23              8                 1   
2     ...      Y                15              8                 3   
3     ...      Y                11              8                 3   
4     ...      Y                12              8                 2   
...   ...    ...               ...            ...               ...   
8815  ...      Y                17              8                 1   
8816  ...      Y                15              8                 0   
8817  ...      Y                20              8                 0   
8818  ...      Y                14              8                 1   
8819  ...      Y                12              8                 0   

     TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                  1.0                      6               1   
1                  6.0                      3               5   
2                  5.0                      2               5   
3                 13.0                      5               8   
4                  9.0                      2               6   
...                ...                    ...             ...   
8815              10.0                      5               3   
8816              10.0                      2               3   
8817               5.0                      4               4   
8818              10.0                      2               9   
8819               NaN                      6              21   

      YearsSinceLastPromotion  YearsWithCurrManager        Year  
0                           0                     0  2015-12-31  
1                           1                     4  2015-12-31  
2                           0                     3  2015-12-31  
3                           7                     5  2015-12-31  
4                           0                     4  2015-12-31  
...                       ...                   ...         ...  
8815                        0                     2  2016-12-31  
8816  

In [8]:
#Renombrar columnna de fecha que aparece como DateSurvey a Year
df_encuesta = df_encuesta.rename(columns={'DateSurvey': 'Year'})

In [9]:
#Revision de columnas
df_encuesta

EmployeeID  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  \
0              1                      3.0              4.0              2.0   
1              2                      3.0              2.0              4.0   
2              3                      2.0              2.0              1.0   
3              4                      4.0              4.0              3.0   
4              5                      4.0              1.0              3.0   
...          ...                      ...              ...              ...   
8815        4406                      4.0              1.0              3.0   
8816        4407                      4.0              4.0              3.0   
8817        4408                      1.0              3.0              3.0   
8818        4409                      4.0              1.0              3.0   
8819        4410                      1.0              3.0              NaN   

            Year  
0     2015-12-31  
1     2015-12-31  
2     2015-12-31  
3     2015-12-31  
4     2015-12-31  
...          ...  
8815  2016-12-31  
8816  2016-12-31  
8817  2016-12-31  
8818  2016-12-31  
8819  2016-12-31  

[8820 rows x 5 columns]

In [9]:
#Renombrar columnna de fecha que aparece como SurveyDate a Year
df_manager = df_manager.rename(columns={'SurveyDate': 'Year'})

In [11]:
#Revision de columnas
df_manager

EmployeeID  JobInvolvement  PerformanceRating        Year
0              1               3                  3  2015-12-31
1              2               2                  4  2015-12-31
2              3               3                  3  2015-12-31
3              4               2                  3  2015-12-31
4              5               3                  3  2015-12-31
...          ...             ...                ...         ...
8815        4406               3                  3  2016-12-31
8816        4407               2                  3  2016-12-31
8817        4408               3                  4  2016-12-31
8818        4409               2                  3  2016-12-31
8819        4410               4                  3  2016-12-31

[8820 rows x 4 columns]

MERGED 1

En el primer merge de la base de datos se priorizó unir dos bases ( General y Encuesta) que tuvieran la misma cantidad de observaciones (8820) para tener claro que el dataframe resultante debía tener 8820 de igual manera, solamente aumentando la cantidad de columnas. La union se realizo con EmployeeID y con Year para garantizar que sean los empleados del mismo año en y el mismo ID

In [10]:
#Union de lo dataframes
df_merged1 = pd.merge(df_general, df_encuesta, on=['EmployeeID', 'Year'], how='outer')


In [13]:
#Revision de columnas
df_merged1

Age     BusinessTravel              Department  DistanceFromHome  \
0      51      Travel_Rarely                   Sales                 6   
1      31  Travel_Frequently  Research & Development                10   
2      32  Travel_Frequently  Research & Development                17   
3      38         Non-Travel  Research & Development                 2   
4      32      Travel_Rarely  Research & Development                10   
...   ...                ...                     ...               ...   
8815   42      Travel_Rarely  Research & Development                 5   
8816   29      Travel_Rarely  Research & Development                 2   
8817   25      Travel_Rarely  Research & Development                25   
8818   42      Travel_Rarely                   Sales                18   
8819   40      Travel_Rarely  Research & Development                28   

      Education EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
0             2  Life Sciences              1           1  Female         1   
1             1  Life Sciences              1           2  Female         1   
2             4          Other              1           3    Male         4   
3             5  Life Sciences              1           4    Male         3   
4             1        Medical              1           5    Male         1   
...         ...            ...            ...         ...     ...       ...   
8815          4        Medical              1        4406  Female         1   
8816          4        Medical              1        4407    Male         1   
8817          2  Life Sciences              1        4408    Male         2   
8818          2        Medical              1        4409    Male         1   
8819          3        Medical              1        4410    Male         2   

      ... StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0     ...                0               1.0                      6   
1     ...                1               6.0                      3   
2     ...                3               5.0                      2   
3     ...                3              13.0                      5   
4     ...                2               9.0                      2   
...   ...              ...               ...                    ...   
8815  ...                1              10.0                      5   
8816  ...                0              10.0                      2   
8817  ...                0               5.0                      4   
8818  ...                1              10.0                      2   
8819  ...                0               0.0                      6   

      YearsAtCompany YearsSinceLastPromotion  YearsWithCurrManager  \
0                  1                       0                     0   
1                  5                       1                     4   
2                  5                       0                     3   
3                  8                       7                     5   
4                  6                       0                     4   
...              ...                     ...                   ...   
8815               3                       0                     2   
8816               3                       0                     2   
8817               4                       1                     2   
8818               9                       7                     8   
8819              21                       3                     9   

            Year  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  
0     2015-12-31                      3.0              4.0              2.0  
1     2015-12-31                      3.0              2.0              4.0  
2     2015-12-31                      2.0              2.0              1.0  
3     2015-12-31                      4.0              4.0              3.0  
4     2015-12-31                      4.0              1.0              3.0  
...   

MERGED 2

En el segundo merged se decide unirlo con la base de datos de manager ya que tambien cuenta con 8820 observaciones y permite una revision facil para confirmar que no presente problemas. Además tambien se vuelven a unir según las columnas de EmployeeID y Year

In [11]:
#creacion del segundo merged donde ya se han consolidado 3 bases de datos 
df_merged2 = pd.merge(df_merged1,df_manager, on=['EmployeeID','Year'],how='outer')

In [29]:
#Revision del segundo merged
df_merged2

Age     BusinessTravel              Department  DistanceFromHome  \
0      51      Travel_Rarely                   Sales                 6   
1      31  Travel_Frequently  Research & Development                10   
2      32  Travel_Frequently  Research & Development                17   
3      38         Non-Travel  Research & Development                 2   
4      32      Travel_Rarely  Research & Development                10   
...   ...                ...                     ...               ...   
8815   42      Travel_Rarely  Research & Development                 5   
8816   29      Travel_Rarely  Research & Development                 2   
8817   25      Travel_Rarely  Research & Development                25   
8818   42      Travel_Rarely                   Sales                18   
8819   40      Travel_Rarely  Research & Development                28   

      Education EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
0             2  Life Sciences              1           1  Female         1   
1             1  Life Sciences              1           2  Female         1   
2             4          Other              1           3    Male         4   
3             5  Life Sciences              1           4    Male         3   
4             1        Medical              1           5    Male         1   
...         ...            ...            ...         ...     ...       ...   
8815          4        Medical              1        4406  Female         1   
8816          4        Medical              1        4407    Male         1   
8817          2  Life Sciences              1        4408    Male         2   
8818          2        Medical              1        4409    Male         1   
8819          3        Medical              1        4410    Male         2   

      ... TrainingTimesLastYear YearsAtCompany  YearsSinceLastPromotion  \
0     ...                     6              1                        0   
1     ...                     3              5                        1   
2     ...                     2              5                        0   
3     ...                     5              8                        7   
4     ...                     2              6                        0   
...   ...                   ...            ...                      ...   
8815  ...                     5              3                        0   
8816  ...                     2              3                        0   
8817  ...                     4              4                        1   
8818  ...                     2              9                        7   
8819  ...                     6             21                        3   

      YearsWithCurrManager        Year  EnvironmentSatisfaction  \
0                        0  2015-12-31                      3.0   
1                        4  2015-12-31                      3.0   
2                        3  2015-12-31                      2.0   
3                        5  2015-12-31                      4.0   
4                        4  2015-12-31                      4.0   
...                    ...         ...                      ...   
8815                     2  2016-12-31                      4.0   
8816                     2  2016-12-31                      4.0   
8817                     2  2016-12-31                      1.0   
8818                     8  2016-12-31                      4.0   
8819                     9  2016-12-31                      1.0   

      JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
0                 4.0              2.0               3                  3  
1                 2.0              4.0               2                  4  
2                 2.0              1.0               3                  3  
3                 4.0              3.0               2                  3  
4                 1.0              3.0               3                  3  
...   

In [12]:
#Se renombra la columna EmployeeID a Col_ID y se pone en la primera columna del dataframe para facilitar la revisión 

col_ID = df_merged2.pop('EmployeeID')

df_merged2.insert(0, 'EmployeeID', col_ID)

In [13]:
#Se renombra la columna Year a Col_Year y se pone en la segunda columna del dataframe para facilitar la revisión 

col_Year = df_merged2.pop('Year')

df_merged2.insert(1, 'Year', col_Year)

In [35]:
#Revision del segundo merged
df_merged2

EmployeeID        Year  Age     BusinessTravel              Department  \
0              1  2015-12-31   51      Travel_Rarely                   Sales   
1              2  2015-12-31   31  Travel_Frequently  Research & Development   
2              3  2015-12-31   32  Travel_Frequently  Research & Development   
3              4  2015-12-31   38         Non-Travel  Research & Development   
4              5  2015-12-31   32      Travel_Rarely  Research & Development   
...          ...         ...  ...                ...                     ...   
8815        4406  2016-12-31   42      Travel_Rarely  Research & Development   
8816        4407  2016-12-31   29      Travel_Rarely  Research & Development   
8817        4408  2016-12-31   25      Travel_Rarely  Research & Development   
8818        4409  2016-12-31   42      Travel_Rarely                   Sales   
8819        4410  2016-12-31   40      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                    6          2  Life Sciences              1  Female  ...   
1                   10          1  Life Sciences              1  Female  ...   
2                   17          4          Other              1    Male  ...   
3                    2          5  Life Sciences              1    Male  ...   
4                   10          1        Medical              1    Male  ...   
...                ...        ...            ...            ...     ...  ...   
8815                 5          4        Medical              1  Female  ...   
8816                 2          4        Medical              1    Male  ...   
8817                25          2  Life Sciences              1    Male  ...   
8818                18          2        Medical              1    Male  ...   
8819                28          3        Medical              1    Male  ...   

      TotalWorkingYears TrainingTimesLastYear YearsAtCompany  \
0                   1.0                     6              1   
1                   6.0                     3              5   
2                   5.0                     2              5   
3                  13.0                     5              8   
4                   9.0                     2              6   
...                 ...                   ...            ...   
8815               10.0                     5              3   
8816               10.0                     2              3   
8817                5.0                     4              4   
8818               10.0                     2              9   
8819                0.0                     6             21   

      YearsSinceLastPromotion  YearsWithCurrManager EnvironmentSatisfaction  \
0                           0                     0                     3.0   
1                           1                     4                     3.0   
2                           0                     3                     2.0   
3                           7                     5                     4.0   
4                           0                     4                     4.0   
...                       ...                   ...                     ...   
8815                        0                     2                     4.0   
8816                        0                     2                     4.0   
8817                        1                     2                     1.0   
8818                        7                     8                     4.0   
8819                        3                     9                     1.0   

      JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
0                 4.0              2.0               3                  3  
1                 2.0              4.0               2                  4  
2                 2.0              1.0               3                  3  
3                 4.0              3.0               2        

In [75]:
'''''
Se utiliza query para revisar los empleaados que se retiraron en el 2015 no deberían estar en 
la base de datos del 2016 porque ellos estarían en el 2014 (la cual no se tiene)
'''
df_merged2.query("EmployeeID == 3995")

EmployeeID  Year  Age BusinessTravel              Department  \
3994        3995  2015   31  Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount Gender  ...  \
3994                 2          3  Life Sciences              1   Male  ...   

      TotalWorkingYears TrainingTimesLastYear YearsAtCompany  \
3994                2.0                     2              1   

      YearsSinceLastPromotion  YearsWithCurrManager EnvironmentSatisfaction  \
3994                        0                     0                     4.0   

      JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
3994              1.0              3.0               3                  3  

[1 rows x 29 columns]

In [14]:
#Cambiar la columna Year a tipo fecha (date time) y se reemplazan los valores de datetime en la columna Year otra vez pero con el nuevo formato
df_merged2['Year'] = pd.to_datetime(df_merged2['Year'])

In [15]:
#Del formato anterior extraemos solo el año ya que es la variable que nos importa
df_merged2['Year'] = df_merged2['Year'].dt.year

In [54]:
#Revision de los cambios efectuados
df_merged2

EmployeeID  Year  Age     BusinessTravel              Department  \
0              1  2015   51      Travel_Rarely                   Sales   
1              2  2015   31  Travel_Frequently  Research & Development   
2              3  2015   32  Travel_Frequently  Research & Development   
3              4  2015   38         Non-Travel  Research & Development   
4              5  2015   32      Travel_Rarely  Research & Development   
...          ...   ...  ...                ...                     ...   
8815        4406  2016   42      Travel_Rarely  Research & Development   
8816        4407  2016   29      Travel_Rarely  Research & Development   
8817        4408  2016   25      Travel_Rarely  Research & Development   
8818        4409  2016   42      Travel_Rarely                   Sales   
8819        4410  2016   40      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                    6          2  Life Sciences              1  Female  ...   
1                   10          1  Life Sciences              1  Female  ...   
2                   17          4          Other              1    Male  ...   
3                    2          5  Life Sciences              1    Male  ...   
4                   10          1        Medical              1    Male  ...   
...                ...        ...            ...            ...     ...  ...   
8815                 5          4        Medical              1  Female  ...   
8816                 2          4        Medical              1    Male  ...   
8817                25          2  Life Sciences              1    Male  ...   
8818                18          2        Medical              1    Male  ...   
8819                28          3        Medical              1    Male  ...   

      TotalWorkingYears TrainingTimesLastYear YearsAtCompany  \
0                   1.0                     6              1   
1                   6.0                     3              5   
2                   5.0                     2              5   
3                  13.0                     5              8   
4                   9.0                     2              6   
...                 ...                   ...            ...   
8815               10.0                     5              3   
8816               10.0                     2              3   
8817                5.0                     4              4   
8818               10.0                     2              9   
8819                0.0                     6             21   

      YearsSinceLastPromotion  YearsWithCurrManager EnvironmentSatisfaction  \
0                           0                     0                     3.0   
1                           1                     4                     3.0   
2                           0                     3                     2.0   
3                           7                     5                     4.0   
4                           0                     4                     4.0   
...                       ...                   ...                     ...   
8815                        0                     2                     4.0   
8816                        0                     2                     4.0   
8817                        1                     2                     1.0   
8818                        7                     8                     4.0   
8819                        3                     9                     1.0   

      JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
0                 4.0              2.0               3                  3  
1                 2.0              4.0               2                  4  
2                 2.0              1.0               3                  3  
3                 4.0              3.0               2                  3  
4                 1.0              3.0               3    

MERGED 3

In [16]:
#Se renombra la columna retirementDate a RTYear
df_retiro = df_retiro.rename(columns={'retirementDate': 'RTYear'})

In [17]:
#Se convierte la columna RTYear a tipo fecha (date time)
df_retiro['RTYear'] = pd.to_datetime(df_retiro['RTYear'])

In [18]:
#Del formato date time extraemos solo el año y la metemos en la nueva columna llamada Year
df_retiro['Year'] = df_retiro['RTYear'].dt.year

In [19]:
#Del formato date time extraemos solo el año y la metemos en la nueva columna llamada RTYear
df_retiro['YearRT'] = df_retiro['RTYear'].dt.year

In [20]:
#Creacion de la nueva columna Month con los datos adquiridos de date time
df_retiro['Month'] = df_retiro['RTYear'].dt.month

In [21]:
#Revision de los cambios
df_retiro

EmployeeID Attrition     RTYear retirementType resignationReason  Year  \
0             2       Yes 2016-10-15    Resignation            Others  2016   
1             7       Yes 2016-07-20    Resignation            Stress  2016   
2            14       Yes 2016-11-08    Resignation            Others  2016   
3            29       Yes 2016-10-15    Resignation            Others  2016   
4            31       Yes 2016-02-08    Resignation            Salary  2016   
..          ...       ...        ...            ...               ...   ...   
706        4382       Yes 2015-04-18    Resignation            Salary  2015   
707        4387       Yes 2015-09-16    Resignation            Others  2015   
708        4389       Yes 2015-05-24    Resignation            Others  2015   
709        4392       Yes 2015-02-26          Fired               NaN  2015   
710        4403       Yes 2015-01-26    Resignation            Stress  2015   

     YearRT  Month  
0      2016     10  
1      2016      7  
2      2016     11  
3      2016     10  
4      2016      2  
..      ...    ...  
706    2015      4  
707    2015      9  
708    2015      5  
709    2015      2  
710    2015      1  

[711 rows x 8 columns]

In [22]:
#Filtramos solamente a un año ( el 2015 no nos interesa de momento) y posteriormente será eliminado 
df_ELIMINAR= df_retiro[df_retiro['Year'] == 2015]

In [23]:
#Se crea una copia del dataframe de retiro y a esta nueva copia se le eliminaron los datos del 2015
df_retiro2 = df_retiro.drop(df_ELIMINAR.index)

In [24]:
#Se elimina la columna RTYear
df_retiro2.drop('RTYear', axis=1, inplace=True)

In [25]:
#Revision de los cambios
df_retiro2

EmployeeID Attrition retirementType resignationReason  Year  YearRT  \
0             2       Yes    Resignation            Others  2016    2016   
1             7       Yes    Resignation            Stress  2016    2016   
2            14       Yes    Resignation            Others  2016    2016   
3            29       Yes    Resignation            Others  2016    2016   
4            31       Yes    Resignation            Salary  2016    2016   
..          ...       ...            ...               ...   ...     ...   
645        3973       Yes    Resignation            Others  2016    2016   
646        3974       Yes    Resignation            Others  2016    2016   
647        3980       Yes    Resignation            Others  2016    2016   
648        3981       Yes    Resignation            Others  2016    2016   
649        3995       Yes    Resignation            Others  2016    2016   

     Month  
0       10  
1        7  
2       11  
3       10  
4        2  
..     ...  
645     11  
646      4  
647      7  
648      2  
649      3  

[650 rows x 7 columns]

In [26]:
#Se resta un año a la columna Year
df_retiro2['Year'] = df_retiro2['Year'] -1

In [27]:
df_retiro2

EmployeeID Attrition retirementType resignationReason  Year  YearRT  \
0             2       Yes    Resignation            Others  2015    2016   
1             7       Yes    Resignation            Stress  2015    2016   
2            14       Yes    Resignation            Others  2015    2016   
3            29       Yes    Resignation            Others  2015    2016   
4            31       Yes    Resignation            Salary  2015    2016   
..          ...       ...            ...               ...   ...     ...   
645        3973       Yes    Resignation            Others  2015    2016   
646        3974       Yes    Resignation            Others  2015    2016   
647        3980       Yes    Resignation            Others  2015    2016   
648        3981       Yes    Resignation            Others  2015    2016   
649        3995       Yes    Resignation            Others  2015    2016   

     Month  
0       10  
1        7  
2       11  
3       10  
4        2  
..     ...  
645     11  
646      4  
647      7  
648      2  
649      3  

[650 rows x 7 columns]

In [28]:
#Creacion del tercer merged. en el cual ya se tienen los datos que nos interesa del dataframe retiro. Adiconmente se unen de acuerdo a EmployeeID y Year
df_merged3 = pd.merge(df_merged2,df_retiro2, on=['EmployeeID','Year'],how='outer')

In [29]:
#DATAFRAME FINAL
df_merged3

EmployeeID  Year  Age     BusinessTravel              Department  \
0              1  2015   51      Travel_Rarely                   Sales   
1              2  2015   31  Travel_Frequently  Research & Development   
2              3  2015   32  Travel_Frequently  Research & Development   
3              4  2015   38         Non-Travel  Research & Development   
4              5  2015   32      Travel_Rarely  Research & Development   
...          ...   ...  ...                ...                     ...   
8815        4406  2016   42      Travel_Rarely  Research & Development   
8816        4407  2016   29      Travel_Rarely  Research & Development   
8817        4408  2016   25      Travel_Rarely  Research & Development   
8818        4409  2016   42      Travel_Rarely                   Sales   
8819        4410  2016   40      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                    6          2  Life Sciences              1  Female  ...   
1                   10          1  Life Sciences              1  Female  ...   
2                   17          4          Other              1    Male  ...   
3                    2          5  Life Sciences              1    Male  ...   
4                   10          1        Medical              1    Male  ...   
...                ...        ...            ...            ...     ...  ...   
8815                 5          4        Medical              1  Female  ...   
8816                 2          4        Medical              1    Male  ...   
8817                25          2  Life Sciences              1    Male  ...   
8818                18          2        Medical              1    Male  ...   
8819                28          3        Medical              1    Male  ...   

      EnvironmentSatisfaction JobSatisfaction WorkLifeBalance  JobInvolvement  \
0                         3.0             4.0             2.0               3   
1                         3.0             2.0             4.0               2   
2                         2.0             2.0             1.0               3   
3                         4.0             4.0             3.0               2   
4                         4.0             1.0             3.0               3   
...                       ...             ...             ...             ...   
8815                      4.0             1.0             3.0               3   
8816                      4.0             4.0             3.0               2   
8817                      1.0             3.0             3.0               3   
8818                      4.0             1.0             3.0               2   
8819                      1.0             3.0             NaN               4   

      PerformanceRating Attrition  retirementType  resignationReason  YearRT  \
0                     3       NaN             NaN                NaN     NaN   
1                     4       Yes     Resignation             Others  2016.0   
2                     3       NaN             NaN                NaN     NaN   
3                     3       NaN             NaN                NaN     NaN   
4                     3       NaN             NaN                NaN     NaN   
...                 ...       ...             ...                ...     ...   
8815                  3       NaN             NaN                NaN     NaN   
8816                  3       NaN             NaN                NaN     NaN   
8817                  4       NaN             NaN                NaN     NaN   
8818                  3       NaN             NaN                NaN     NaN   
8819                  3       NaN             NaN                NaN     NaN   

      Month  
0       NaN  
1      10.0  
2       NaN  
3       NaN  
4       NaN  
...     ...  
8815    NaN  
8816    NaN  
8817    NaN  
8818    NaN  
8819    NaN  

[8820 rows x 34 columns]

IDENTIFICACIÓN E IMPUTACIÓN DE NULOS

In [30]:
df_merged3[df_merged3['Attrition']=='Yes']

EmployeeID  Year  Age     BusinessTravel              Department  \
1              2  2015   31  Travel_Frequently  Research & Development   
6              7  2015   28      Travel_Rarely  Research & Development   
13            14  2015   47         Non-Travel  Research & Development   
28            29  2015   44  Travel_Frequently  Research & Development   
30            31  2015   26      Travel_Rarely  Research & Development   
...          ...   ...  ...                ...                     ...   
3972        3973  2015   34  Travel_Frequently  Research & Development   
3973        3974  2015   37      Travel_Rarely  Research & Development   
3979        3980  2015   26      Travel_Rarely         Human Resources   
3980        3981  2015   46      Travel_Rarely  Research & Development   
3994        3995  2015   31      Travel_Rarely  Research & Development   

      DistanceFromHome  Education    EducationField  EmployeeCount  Gender  \
1                   10          1     Life Sciences              1  Female   
6                   11          2           Medical              1    Male   
13                   1          1           Medical              1    Male   
28                   1          2           Medical              1    Male   
30                   4          3           Medical              1    Male   
...                ...        ...               ...            ...     ...   
3972                 2          1     Life Sciences              1    Male   
3973                24          3     Life Sciences              1    Male   
3979                 2          1  Technical Degree              1  Female   
3980                10          3           Medical              1    Male   
3994                 2          3     Life Sciences              1    Male   

      ...  EnvironmentSatisfaction JobSatisfaction WorkLifeBalance  \
1     ...                      3.0             2.0             4.0   
6     ...                      1.0             3.0             1.0   
13    ...                      1.0             2.0             2.0   
28    ...                      4.0             3.0             1.0   
30    ...                      1.0             2.0             3.0   
...   ...                      ...             ...             ...   
3972  ...                      4.0             1.0             2.0   
3973  ...                      4.0             1.0             1.0   
3979  ...                      4.0             2.0             3.0   
3980  ...                      4.0             2.0             4.0   
3994  ...                      4.0             1.0             3.0   

      JobInvolvement  PerformanceRating Attrition  retirementType  \
1                  2                  4       Yes     Resignation   
6                  3                  4       Yes     Resignation   
13                 2                  3       Yes     Resignation   
28                 2                  3       Yes     Resignation   
30                 2                  3       Yes     Resignation   
...              ...                ...       ...             ...   
3972               4                  4       Yes     Resignation   
3973               3                  3       Yes     Resignation   
3979               2                  3       Yes     Resignation   
3980               3                  3       Yes     Resignation   
3994               3                  3       Yes     Resignation   

      resignationReason  YearRT  Month  
1                Others  2016.0   10.0  
6                Stress  2016.0    7.0  
13               Others  2016.0   11.0  
28               Others  2016.0   10.0  
30               Salary  2016.0    2.0  
...                 ...     ...    ...  
3972             Others  2016.0   11.0  
3973             Others  2016.0    4.0  
3979             Others  2016.0    7.0  
3980             Others  2016.0    2.0  
3994             Others  2016.0    3.0  

[650 rows x 34 co

In [31]:
# Columnas que contienen valores nulos
columns_with_nulls = df_merged3.columns[df_merged3.isnull().any()]
columns_with_nulls

Index(['NumCompaniesWorked', 'TotalWorkingYears', 'EnvironmentSatisfaction',
       'JobSatisfaction', 'WorkLifeBalance', 'Attrition', 'retirementType',
       'resignationReason', 'YearRT', 'Month'],
      dtype='object')

In [32]:
df_merged3['WorkLifeBalance'].value_counts()

WorkLifeBalance
3.0    5320
2.0    2038
4.0     908
1.0     478
Name: count, dtype: int64

Se identifica que los datos nulos de algunas variables pueden ser reemplazadas con la mediana: 'NumCompaniesWorked', 'TotalWorkingYears', 'EnvironmentSatisfaction', 'JobSatisfaction' y 'WorkLifeBalance'.

In [33]:
# Imputación por la mediana
df_merged3["NumCompaniesWorked"]=df_merged3["NumCompaniesWorked"].apply(lambda x: x if not pd.isnull(x) else int(df_merged3["NumCompaniesWorked"].median()))
df_merged3["TotalWorkingYears"]=df_merged3["TotalWorkingYears"].apply(lambda x: x if not pd.isnull(x) else int(df_merged3["TotalWorkingYears"].median()))
df_merged3["EnvironmentSatisfaction"]=df_merged3["EnvironmentSatisfaction"].apply(lambda x: x if not pd.isnull(x) else int(df_merged3["EnvironmentSatisfaction"].median()))
df_merged3["JobSatisfaction"]=df_merged3["JobSatisfaction"].apply(lambda x: x if not pd.isnull(x) else int(df_merged3["JobSatisfaction"].median()))
df_merged3["WorkLifeBalance"]=df_merged3["WorkLifeBalance"].apply(lambda x: x if not pd.isnull(x) else int(df_merged3["WorkLifeBalance"].median()))

In [34]:
# Columnas que contienen valores nulos
columns_with_nulls = df_merged3.columns[df_merged3.isnull().any()]
columns_with_nulls

Index(['Attrition', 'retirementType', 'resignationReason', 'YearRT', 'Month'], dtype='object')

Y en el caso de las variables restantes (relacionadas con los datos de retiros) se entiende que las observaciones con datos nulos son los empleados que no se retiran. Por ende, la imputación de dichas datos debe ser por 'No aplica' o por valores de ese estilo.

In [35]:
# En el caso de 'Attrition' se reemplaza los nulos por 'No' dado que representa los empleados que no se retiran
df_merged3['Attrition'] = df_merged3['Attrition'].fillna('No')

In [36]:
# Además, con el objetivo de convertir la variable 'Attrition' en la variable objetivo más adelante, se reempaza 'Yes' por 1 y 'No' por 0 
df_merged3['Attrition'] = df_merged3['Attrition'].replace({'Yes': 1, 'No': 0})

In [37]:
# En el caso de las variables restantes ('retirementType', 'resignationReason', 'YearRt' y 'Month') se reemplaza los nulos por 'Not Applicable'
df_merged3['retirementType'] = df_merged3['retirementType'].fillna('Not Applicable')
df_merged3['resignationReason'] = df_merged3['resignationReason'].fillna('Not Applicable')
df_merged3['YearRT'] = df_merged3['YearRT'].fillna('Not Applicable')
df_merged3['Month'] = df_merged3['Month'].fillna('Not Applicable')

In [38]:
df_merged3.columns

Index(['EmployeeID', 'Year', 'Age', 'BusinessTravel', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'Gender', 'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager',
       'EnvironmentSatisfaction', 'JobSatisfaction', 'WorkLifeBalance',
       'JobInvolvement', 'PerformanceRating', 'Attrition', 'retirementType',
       'resignationReason', 'YearRT', 'Month'],
      dtype='object')

In [39]:
# Se realiza el cambio de tipología de algunas variables
df_merged3['TotalWorkingYears'] = df_merged3['TotalWorkingYears'].astype(int)
df_merged3['EnvironmentSatisfaction'] = df_merged3['EnvironmentSatisfaction'].astype(int)
df_merged3['JobSatisfaction'] = df_merged3['JobSatisfaction'].astype(int)
df_merged3['WorkLifeBalance'] = df_merged3['WorkLifeBalance'].astype(int)
df_merged3['NumCompaniesWorked'] = df_merged3['NumCompaniesWorked'].astype(int)

In [40]:
df_merged3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8820 entries, 0 to 8819
Data columns (total 34 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   EmployeeID               8820 non-null   int64 
 1   Year                     8820 non-null   int32 
 2   Age                      8820 non-null   int64 
 3   BusinessTravel           8820 non-null   object
 4   Department               8820 non-null   object
 5   DistanceFromHome         8820 non-null   int64 
 6   Education                8820 non-null   int64 
 7   EducationField           8820 non-null   object
 8   EmployeeCount            8820 non-null   int64 
 9   Gender                   8820 non-null   object
 10  JobLevel                 8820 non-null   int64 
 11  JobRole                  8820 non-null   object
 12  MaritalStatus            8820 non-null   object
 13  MonthlyIncome            8820 non-null   int64 
 14  NumCompaniesWorked       8820 non-null  

In [41]:
df_merged3.head(8)

EmployeeID  Year  Age     BusinessTravel              Department  \
0           1  2015   51      Travel_Rarely                   Sales   
1           2  2015   31  Travel_Frequently  Research & Development   
2           3  2015   32  Travel_Frequently  Research & Development   
3           4  2015   38         Non-Travel  Research & Development   
4           5  2015   32      Travel_Rarely  Research & Development   
5           6  2015   46      Travel_Rarely  Research & Development   
6           7  2015   28      Travel_Rarely  Research & Development   
7           8  2015   29      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                 6          2  Life Sciences              1  Female  ...   
1                10          1  Life Sciences              1  Female  ...   
2                17          4          Other              1    Male  ...   
3                 2          5  Life Sciences              1    Male  ...   
4                10          1        Medical              1    Male  ...   
5                 8          3  Life Sciences              1  Female  ...   
6                11          2        Medical              1    Male  ...   
7                18          3  Life Sciences              1    Male  ...   

   EnvironmentSatisfaction JobSatisfaction WorkLifeBalance  JobInvolvement  \
0                        3               4               2               3   
1                        3               2               4               2   
2                        2               2               1               3   
3                        4               4               3               2   
4                        4               1               3               3   
5                        3               2               2               3   
6                        1               3               1               3   
7                        1               2               3               3   

   PerformanceRating Attrition  retirementType  resignationReason  \
0                  3         0  Not Applicable     Not Applicable   
1                  4         1     Resignation             Others   
2                  3         0  Not Applicable     Not Applicable   
3                  3         0  Not Applicable     Not Applicable   
4                  3         0  Not Applicable     Not Applicable   
5                  3         0  Not Applicable     Not Applicable   
6                  4         1     Resignation             Stress   
7                  4         0  Not Applicable     Not Applicable   

           YearRT           Month  
0  Not Applicable  Not Applicable  
1          2016.0            10.0  
2  Not Applicable  Not Applicable  
3  Not Applicable  Not Applicable  
4  Not Applicable  Not Applicable  
5  Not Applicable  Not Applicable  
6          2016.0             7.0  
7  Not Applicable  Not Applicable  

[8 rows x 34 columns]

In [42]:
df_merged3['JobRole'].value_counts()

JobRole
Sales Executive              1956
Research Scientist           1752
Laboratory Technician        1554
Manufacturing Director        870
Healthcare Representative     786
Manager                       612
Sales Representative          498
Research Director             480
Human Resources               312
Name: count, dtype: int64

In [43]:
# Datos completos de solo los empleados que se retiran a usar en dashboard
df_only_retired = df_merged3[df_merged3['Attrition']==1]
df_only_retired

EmployeeID  Year  Age     BusinessTravel              Department  \
1              2  2015   31  Travel_Frequently  Research & Development   
6              7  2015   28      Travel_Rarely  Research & Development   
13            14  2015   47         Non-Travel  Research & Development   
28            29  2015   44  Travel_Frequently  Research & Development   
30            31  2015   26      Travel_Rarely  Research & Development   
...          ...   ...  ...                ...                     ...   
3972        3973  2015   34  Travel_Frequently  Research & Development   
3973        3974  2015   37      Travel_Rarely  Research & Development   
3979        3980  2015   26      Travel_Rarely         Human Resources   
3980        3981  2015   46      Travel_Rarely  Research & Development   
3994        3995  2015   31      Travel_Rarely  Research & Development   

      DistanceFromHome  Education    EducationField  EmployeeCount  Gender  \
1                   10          1     Life Sciences              1  Female   
6                   11          2           Medical              1    Male   
13                   1          1           Medical              1    Male   
28                   1          2           Medical              1    Male   
30                   4          3           Medical              1    Male   
...                ...        ...               ...            ...     ...   
3972                 2          1     Life Sciences              1    Male   
3973                24          3     Life Sciences              1    Male   
3979                 2          1  Technical Degree              1  Female   
3980                10          3           Medical              1    Male   
3994                 2          3     Life Sciences              1    Male   

      ...  EnvironmentSatisfaction JobSatisfaction WorkLifeBalance  \
1     ...                        3               2               4   
6     ...                        1               3               1   
13    ...                        1               2               2   
28    ...                        4               3               1   
30    ...                        1               2               3   
...   ...                      ...             ...             ...   
3972  ...                        4               1               2   
3973  ...                        4               1               1   
3979  ...                        4               2               3   
3980  ...                        4               2               4   
3994  ...                        4               1               3   

      JobInvolvement  PerformanceRating Attrition  retirementType  \
1                  2                  4         1     Resignation   
6                  3                  4         1     Resignation   
13                 2                  3         1     Resignation   
28                 2                  3         1     Resignation   
30                 2                  3         1     Resignation   
...              ...                ...       ...             ...   
3972               4                  4         1     Resignation   
3973               3                  3         1     Resignation   
3979               2                  3         1     Resignation   
3980               3                  3         1     Resignation   
3994               3                  3         1     Resignation   

      resignationReason  YearRT  Month  
1                Others  2016.0   10.0  
6                Stress  2016.0    7.0  
13               Others  2016.0   11.0  
28               Others  2016.0   10.0  
30               Salary  2016.0    2.0  
...                 ...     ...    ...  
3972             Others  2016.0   11.0  
3973             Others  2016.0    4.0  
3979             Others  2016.0    7.0  
3980             Others  2016.0    2.0  
3994             Others  2016.0    3.0  

[650 rows x 34 co

In [44]:
# Exportación de los datos
df_only_retired.to_csv('only_retired.csv', index=False)

In [45]:
# Moviendo la columna 'Attrition' 2 posiciones a la derecha
cols = df_merged3.columns.tolist()
cols.insert(cols.index('Attrition')+2, cols.pop(cols.index('Attrition')))
df_merged3 = df_merged3[cols]

In [46]:
df_merged3.head(8)

EmployeeID  Year  Age     BusinessTravel              Department  \
0           1  2015   51      Travel_Rarely                   Sales   
1           2  2015   31  Travel_Frequently  Research & Development   
2           3  2015   32  Travel_Frequently  Research & Development   
3           4  2015   38         Non-Travel  Research & Development   
4           5  2015   32      Travel_Rarely  Research & Development   
5           6  2015   46      Travel_Rarely  Research & Development   
6           7  2015   28      Travel_Rarely  Research & Development   
7           8  2015   29      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                 6          2  Life Sciences              1  Female  ...   
1                10          1  Life Sciences              1  Female  ...   
2                17          4          Other              1    Male  ...   
3                 2          5  Life Sciences              1    Male  ...   
4                10          1        Medical              1    Male  ...   
5                 8          3  Life Sciences              1  Female  ...   
6                11          2        Medical              1    Male  ...   
7                18          3  Life Sciences              1    Male  ...   

   EnvironmentSatisfaction JobSatisfaction WorkLifeBalance  JobInvolvement  \
0                        3               4               2               3   
1                        3               2               4               2   
2                        2               2               1               3   
3                        4               4               3               2   
4                        4               1               3               3   
5                        3               2               2               3   
6                        1               3               1               3   
7                        1               2               3               3   

   PerformanceRating  retirementType  resignationReason  Attrition  \
0                  3  Not Applicable     Not Applicable          0   
1                  4     Resignation             Others          1   
2                  3  Not Applicable     Not Applicable          0   
3                  3  Not Applicable     Not Applicable          0   
4                  3  Not Applicable     Not Applicable          0   
5                  3  Not Applicable     Not Applicable          0   
6                  4     Resignation             Stress          1   
7                  4  Not Applicable     Not Applicable          0   

           YearRT           Month  
0  Not Applicable  Not Applicable  
1          2016.0            10.0  
2  Not Applicable  Not Applicable  
3  Not Applicable  Not Applicable  
4  Not Applicable  Not Applicable  
5  Not Applicable  Not Applicable  
6          2016.0             7.0  
7  Not Applicable  Not Applicable  

[8 rows x 34 columns]

In [47]:
# Datos completos de todos los empleados solo hasta la variable 'Attrition', dejando por fuera el mes y el año del retiro

# Obtener las etiquetas de todas las columnas excepto las dos últimas
columnas_a_mantener = df_merged3.columns[:-2]

# Crear un nuevo DataFrame con las columnas seleccionadas
df_entire_data = df_merged3[columnas_a_mantener]
df_entire_data

EmployeeID  Year  Age     BusinessTravel              Department  \
0              1  2015   51      Travel_Rarely                   Sales   
1              2  2015   31  Travel_Frequently  Research & Development   
2              3  2015   32  Travel_Frequently  Research & Development   
3              4  2015   38         Non-Travel  Research & Development   
4              5  2015   32      Travel_Rarely  Research & Development   
...          ...   ...  ...                ...                     ...   
8815        4406  2016   42      Travel_Rarely  Research & Development   
8816        4407  2016   29      Travel_Rarely  Research & Development   
8817        4408  2016   25      Travel_Rarely  Research & Development   
8818        4409  2016   42      Travel_Rarely                   Sales   
8819        4410  2016   40      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  Gender  ...  \
0                    6          2  Life Sciences              1  Female  ...   
1                   10          1  Life Sciences              1  Female  ...   
2                   17          4          Other              1    Male  ...   
3                    2          5  Life Sciences              1    Male  ...   
4                   10          1        Medical              1    Male  ...   
...                ...        ...            ...            ...     ...  ...   
8815                 5          4        Medical              1  Female  ...   
8816                 2          4        Medical              1    Male  ...   
8817                25          2  Life Sciences              1    Male  ...   
8818                18          2        Medical              1    Male  ...   
8819                28          3        Medical              1    Male  ...   

      YearsSinceLastPromotion YearsWithCurrManager EnvironmentSatisfaction  \
0                           0                    0                       3   
1                           1                    4                       3   
2                           0                    3                       2   
3                           7                    5                       4   
4                           0                    4                       4   
...                       ...                  ...                     ...   
8815                        0                    2                       4   
8816                        0                    2                       4   
8817                        1                    2                       1   
8818                        7                    8                       4   
8819                        3                    9                       1   

      JobSatisfaction  WorkLifeBalance JobInvolvement  PerformanceRating  \
0                   4                2              3                  3   
1                   2                4              2                  4   
2                   2                1              3                  3   
3                   4                3              2                  3   
4                   1                3              3                  3   
...               ...              ...            ...                ...   
8815                1                3              3                  3   
8816                4                3              2                  3   
8817                3                3              3                  4   
8818                1                3              2                  3   
8819                3                3              4                  3   

      retirementType  resignationReason  Attrition  
0     Not Applicable     Not Applicable          0  
1        Resignation             Others          1  
2     Not Applicable     Not Applicable          0  
3     Not Applicable     Not Applicable          0  
4     Not Applicable     Not Applicable  

In [48]:
df_entire_data.to_csv('entire_data.csv', index=False)